source/agora/node/Ledger.d

```diff
@@ -1260,6 +1275,11 @@ public class Ledger
                 missing_validators_lower_bound ~= cast(uint)idx;
         }
 
+        // trying to retrieve preimages that - based on the consensus data -
+        // was shared with other nodes
+        foreach (const validator_ind; setDifference(missing_validators_lower_bound, missing_validators))
+            this.enrolls_keys_for_unknown_preimages.put(keys[validator_ind]);
+
```

Comment:
reset is in NetworkManager.retrievePreimages()
```D
        () @trusted {enroll_keys.clear();}();
```

Reply:
'However, with this addition, it no longer is, and there's an assumption that it will only be called once per' 

it is a Set!Hash - there is no assumption

Reply:
'This method was not originally modifying any class member, which means it was idempotent.'

suggestion where it should be?

Reply:
Discussed offline, we're going with this approach.